Refactor more LDAP code into separate package

alistanis opened the pull request:
This PR moves a few more things LDAP related into the new auth/windows package.

I don't think that the mutex from the windows desktop service conflicts with these changes, but I'm not 100% sure if I've broken consistency here, so I'd appreciate some eyes on that in particular.

This PR is gated by https://github.com/gravitational/teleport/pull/18299, which needs to go in first.

alistanis committed:
Refactor more LDAP code into separate package

zmb3 commented on lib/srv/db/sqlserver/kinit/kinit.go:
```diff
@@ -0,0 +1,210 @@
+
+// CreateOrAppendCredentialsCache creates or appends to an existing credentials cache.
+func (k *CommandLineKInit) CreateOrAppendCredentialsCache(ctx context.Context) error {
+
+	tmp := filepath.Join(os.TempDir(), "kinit")
+	err := os.MkdirAll(tmp, 0664)
+	if err != nil {
+		return trace.Wrap(err)
+	}
```
Use `os.MkdirTemp` instead?

zmb3 commented on lib/srv/db/sqlserver/kinit/kinit.go:
```diff
@@ -0,0 +1,210 @@
+
+	err = os.WriteFile(keyPath, keyPEM, 0644)
+	if err != nil {
+		return trace.Wrap(err)
```
Are you okay with leaving files behind in the temp dir if one of these writes failed? I would try to clean them up.

zmb3 commented on lib/srv/db/sqlserver/kinit/kinit.go:
```diff
@@ -0,0 +1,210 @@
+
+	if !bytes.Contains(data, []byte(fmt.Sprintf(`Storing %s@%s -> `, k.userName, strings.ToUpper(k.realmName)))) {
+		k.log.Debug(string(data))
+		return trace.Wrap(fmt.Errorf("unable to store credentials for user: %s, output: %s", k.userName, string(data)))
```
Would `trace.AccessDenied` be appropriate here?

zmb3 commented on lib/srv/db/sqlserver/kinit/kinit.go:
```diff
@@ -0,0 +1,210 @@
+	}
+
+	if caCert == nil {
+		return trace.Wrap(errors.New("no certificate authority was found in userCA active keys"))
```
We should probably update the trace package to make this easier. Maybe add something like `trace.New()`..

zmb3 commented on lib/srv/db/sqlserver/kinit/kinit.go:
```diff
@@ -0,0 +1,210 @@
+// krb5ConfigString returns a config suitable for a kdc
+func (k *CommandLineKInit) krb5ConfigString() string {
+	return fmt.Sprintf(`[libdefaults]
+ default_realm = %s
```
Totally up to you, but sometimes when things like this start to get too big I prefer the use of `text/template` instead of `Sprintf` because it's easier to see which `%s` goes where.

This particular example is not too bad, so I don't feel strongly - I'll leave it to you to decide. It would look something like this when done:

```
[libdefaults]
default_realm = {{ .RealmName }}
rdns = false

[realms]
{{ .RealmName }} = {
  kdc = {{ .KDCHostName }}
  ...
```

zmb3 commented on lib/srv/desktop/windows_server.go:
```diff
@@ -106,9 +95,7 @@ type WindowsService struct {
-	mu              sync.Mutex // mu protects the fields that follow
-	ldapInitialized bool
-	ldapCertRenew   *time.Timer
+	mu sync.Mutex // mu protects the fields that follow
```
I don't think you need this at all anymore - it was only to protect `ldapInitialized` and `ldapCertRenew`.

alistanis replied:
This was my biggest question here - it didn't look like it was used for anything else but I wasn't 100% sure. Thanks!

zmb3 commented:
@ibeckermayer would you mind taking this for a test drive before we merge it just so we don't break desktop access again? 😄

ibeckermayer reviewed (commented):
It's working for me in the present state, tag me once you've addressed feedback and I'll give it another run just to confirm.

alistanis commented:
> It's working for me in the present state, tag me once you've addressed feedback and I'll give it another run just to confirm.

Thanks @ibeckermayer, I appreciate it. I need to get the other PR ahead of this one in first, then I'll rebase and address this one and give you a shout.

smallinsky commented on lib/auth/windows/windows.go:
```diff
@@ -103,6 +103,12 @@ type AuthInterface interface {
+}
+
+// CachedAuthInterface is a subset
+type CachedAuthInterface interface {
```
Seems to be unused.

smallinsky commented on api/types/database.go:
```diff
@@ -634,6 +636,17 @@ func (d *DatabaseV3) CheckAndSetDefaults() error {
+		// cannot use defaults here due to import cycle
+		if d.Spec.Protocol == "sqlserver" {
+			if d.Spec.AD.Krb5File == "" {
+				d.Spec.AD.Krb5File = filepath.Join("etc", "krb5.conf")
```
it will return "etc/krb5.conf" that not looks valid if process dir is not '/'

smallinsky commented on lib/srv/db/server.go:
```diff
@@ -71,6 +72,8 @@ type Config struct {
 	// AccessPoint is a caching client connected to the Auth Server.
 	AccessPoint auth.DatabaseAccessPoint
+	// LDAPConfig is an LDAP configuration for connecting to SQL Server with x509 authentication
+	LDAPConfig windows.LDAPConfig
```
Where this  `LDAPConfig`  is set for the db.Config  ?

smallinsky commented on lib/srv/db/server.go:
```diff
@@ -358,6 +435,13 @@ func (s *Server) startDatabase(ctx context.Context, database types.Database) err
 	}
+
+	if database.GetProtocol() == defaults.ProtocolSQLServer {
+		if err := s.startSQLServerLDAP(ctx, database); err != nil {
+			return trace.Wrap(err)
+		}
```
Why do we want always to start SQLServerLDAP ? For the old configuration when the `database.GetAD().LDAPCert` is empty the `startSQLServerLDAP` will returns an error and block the service from running.

smallinsky commented on lib/srv/db/server.go:
```diff
@@ -326,6 +344,65 @@ func New(ctx context.Context, config Config) (*Server, error) {
+		serverName = ad.ServerName
+	}
+
+	s.caClients[database.GetName()] = windows.NewCertificateStoreClient(windows.CertificateStoreConfig{
```
Does it make sense to use domains as a key  instead of database name  and detect the situation when the initializeLDAP is already running for particular domain ?

smallinsky commented on lib/srv/db/server.go:
```diff
@@ -358,6 +435,13 @@ func (s *Server) startDatabase(ctx context.Context, database types.Database) err
 	}
+
+	if database.GetProtocol() == defaults.ProtocolSQLServer {
+		if err := s.startSQLServerLDAP(ctx, database); err != nil {
```
@alistanis 
The startSQLServerLDAP functions calls
`startSQLServerLDAP-> InitializeLDAP-> tlsConfigForLDAP-> GenerateCredentials` 

where [GenerateCredentials](https://github.com/gravitational/teleport/blob/6d74ef49fdf079c33c7cd1e0f9a3eb5c354f9009/lib/auth/windows/windows.go#L115)
does the `authClient.GenerateWindowsDesktopCert` call 

How does the DB service `startSQLServerLDAP` call works if `GenerateWindowsDesktopCert` endpoint allows to be invoked only by `RoleWindowsDesktop` role [link](https://github.com/gravitational/teleport/blob/dd67a49ab5035e51847a3fc8934f53b5abe25a6a/lib/auth/auth_with_roles.go#L5060) ?

alistanis committed:
Do not run ldap init when LDAPCert is empty

zmb3 commented:
Closing due to inactivity
